### PR TITLE
Pre-bind sockets on creation

### LINKED ISSUE #31: GELF Benchmarks are slow vs UDP
Normally I would pair this issue with a pull request but your guys requirement to sign a disclosure and waiver is something I can't do.  That 'release form' is completely different than the license you've attached to this repo so I don't know what to believe.

That all said, I'm going to list out what I would fix if this was a normal open source project in hopes that you agree and will implement the change.

First, the problem.  GELF was slowing down our servers, which seems odd especially considering GELF uses UDP under the hood which is a 'fire and forget' protocol.  I decided to Benchmark UDP vs. GELF's implementation of UDP.

###The Benchmark results:

1. Benchmark/ip:

```ruby
Calculating -------------------------------------
                GELF       275 i/100ms
                 UDP     10343 i/100ms
-------------------------------------------------
                GELF     2760.6 (±3.9%) i/s -      14025 in   5.088188s
                 UDP    97956.3 (±21.8%) i/s -     465435 in   5.064302s
```

2. Benchmark.bm

```ruby
                      user     system      total        real
GELF             16.400000  10.830000  27.230000 ( 39.284368)
UDP               0.300000   0.560000   0.860000 (  0.871986)
```

Staggering really - UDP can log 90_000-100_000 msgs / second, GELF's implementation can do 2500-3000 msgs / second.

The benchmark script looks like this:

```ruby
require 'gelf'
require 'benchmark/ips'

# GELF set up
glogger = GELF::Logger.new("localhost", 5352)

# UDP set up
require 'socket'
s = UDPSocket.new
s.connect('127.0.0.1', 1234)

Benchmark.ips do |x|
  x.report("GELF") { glogger.info 1 }
  x.report("UDP")  { s.send("#{1}", 0) }
end

require 'benchmark'

Benchmark.bm(15) do |x|
  x.report("GELF") { 100_000.times { |i| glogger.info i } }
  x.report("UDP")  { 100_000.times { |i| s.send("#{i}", 0) } }
end
```

You'll need to run a UDP server in a separate process for those to work:

```ruby
require 'socket'
s = UDPSocket.new
s.bind(nil, 1234)
loop do
  text, sender = s.recvfrom(16)
end
```

So what's different?  Well it's really a small change.  In the GELF gem, we never open the connection until we call the `send` method with arguments.

### PR DESCRIPTION
**Pull request by markglenn**
When sending a datagram, non-bound sockets require a DNS lookup.
Instead, bind the socket to the given addresses when creating the
socket.

Also, lazily create the socket connections, which allows the
notify methods to capture the exceptions as expected.  If the sockets
are created and bound at initialize, an exception would occur then.

Lastly, allow resetting the connections if the addresses are changed.

Fixes #31

### CODE REVIEW

File: lib/gelf/ruby_sender.rb [hunk 1]
```diff
@@ -4,21 +4,44 @@ class RubyUdpSender
...
+
+      @addresses = addresses
+    end
+
+    private
+
+    # Get UDPSockets that are prebound to the given addresses
+    def sockets
+      @sockets ||= addresses.map do |address|
+        host, port = address
```
Comment (matugm) [inline lib/gelf/ruby_sender.rb]: I think you don't need this, you could try:

```ruby
sockets ||= addresses.map do |host, port|  
  socket = UDPSocket.new
  socket.connect(host, port)
  socket
end
```

File: lib/gelf/ruby_sender.rb [hunk 2]
```diff
@@ -4,21 +4,44 @@ class RubyUdpSender
     attr_accessor :addresses
 
     def initialize(addresses)
-      @addresses = addresses
+      self.addresses = addresses
```
Comment (matugm) [inline lib/gelf/ruby_sender.rb]: Why did you change from @ to self.?

**Comment by markglenn** [comment]
Both of these concerns are valid and would be fixed, however, I don't think this will be merged in.  See https://github.com/Graylog2/gelf-rb/issues/31#issuecomment-103658710

**Comment by markglenn** [comment]
I'm closing this since I don't believe this is the right technique because it's not safe for networks that allows the Graylog server to change IP addresses.  If you believe I'm wrong, please re-open.
